src/geometry/rtree.cpp

```diff
@@ -0,0 +1,290 @@
+  return std::visit(
+      [](auto&& geom) {
+        if constexpr (std::is_same_v<point_type, std::decay_t<decltype(geom)>>) {
+          constexpr auto EPSILON = 1e-10;
+          auto p1 = point_type{bg::get<0>(geom) - EPSILON, bg::get<1>(geom) - EPSILON};
+          auto p2 = point_type{bg::get<0>(geom) + EPSILON, bg::get<1>(geom) + EPSILON};
+          return rect_type{p1, p2};
+        } else {
+          return bg::return_envelope<rect_type>(geom);
+        }
+      },
+      geom);
+}
+template <typename coord_system>
+constexpr auto RTree<coord_system>::make_doc(geom_type const& geom, t_docId id) -> doc_type {
+  return doc_type{make_mbr(geom), id};
+}
+template <typename coord_system>
+constexpr auto RTree<coord_system>::get_rect(doc_type const& doc) -> rect_type {
+  return doc.first;
+}
+template <typename coord_system>
+constexpr auto RTree<coord_system>::get_id(doc_type const& doc) -> t_docId {
+  return doc.second;
+}
+
+template <typename coord_system>
+auto RTree<coord_system>::lookup(t_docId id) const
+    -> std::optional<std::reference_wrapper<const geom_type>> {
+  if (auto it = docLookup_.find(id); it != docLookup_.end()) {
+    return it->second;
+  }
+  return std::nullopt;
+}
+template <typename coord_system>
+auto RTree<coord_system>::lookup(doc_type const& doc) const
+    -> std::optional<std::reference_wrapper<const geom_type>> {
+  return lookup(get_id(doc));
+}
+
+template <typename coord_system>
+auto RTree<coord_system>::from_wkt(std::string_view wkt) const -> geom_type {
+  geom_type geom{};
+  if (wkt.starts_with("POI")) {
+    geom = bg::from_wkt<point_type>(wkt.data());
+  } else if (wkt.starts_with("POL")) {
+    geom = bg::from_wkt<poly_type>(wkt.data());
+  } else {
+    throw std::runtime_error{"unknown geometry type"};
+  }
+  std::visit(
+      [](auto&& geom) {
+        if (bg::is_empty(geom)) {
+          throw std::runtime_error{"attempting to create empty geometry"};
+        }
+        if (!bg::is_valid(geom)) {  // TODO: GEOMETRY - add flag to allow user to ascertain
+                                    // validity of input
+          throw std::runtime_error{"invalid geometry"};
+        }
+      },
+      geom);
+  return geom;
+}
+
+template <typename coord_system>
+constexpr auto geometry_reporter = [](auto&& geom) {
+  using point_type = typename RTree<coord_system>::point_type;
+  if constexpr (std::is_same_v<point_type, std::decay_t<decltype(geom)>>) {
+    return 0ul;
+  } else {
+    auto const& inners = geom.inners();
+    auto outer_size = geom.outer().get_allocator().report();
+    return std::accumulate(
+        inners.begin(), inners.end(), outer_size,
+        [](std::size_t acc, auto&& hole) { return acc + hole.get_allocator().report(); });
+  }
+};
+
+template <typename coord_system>
+void RTree<coord_system>::insert(geom_type const& geom, t_docId id) {
+  rtree_.insert(make_doc(geom, id));
+  docLookup_.insert(LUT_value_type{id, geom});
```

Comment:
The lookup is using the same allocator, so it would be tracked by it?

Reply:
the lookup is only tracking its own allocations. not allocations done by the objects it stores.